### ISSUE TITLE
handling secrets with gitlabform

### ISSUE THREAD

**Issue by weakcamel** [egnyte/gitlabform#57]
GitlabForm can be succesfully used to provision Gitlab projects with secrets, like passwords, API keys etc.

This is extremely helpful. It also means that with the current design, those secrets need to be hardcoded inside the `config.yml` file

It be great if sensitive data could be passed from external sources. Reasons:

- security; developers in the company may be privy to seeing general configuration of Gitlab so as a reference, so that they know what to expect in a specific project.. yet at the same time, secrets shouldn't be that easily visible
- single source of truth: users of GitlabForm may be keeping their secrets in various safe places and apps: [Vault](https://www.vaultproject.io/), [LastPass](https://www.lastpass.com/), other password managers, even in an encrypted text file or a sticky note (in a safe, obviously :) )
- additionallly (nice to have), the secrets shouldn't be visible in visible in plain text in GitlabForm logs

At the moment the only workaround for that that I can think of would be to keep have `config.yml` file as a template to be filled in at runtime (e.g. with https://pypi.org/project/j2cli/ or https://pypi.org/project/jinja2-vault/). This however:
- complicates the flow as you'd need to re-generate the config each time you want to apply it, lint it (see https://github.com/adrienverge/yamllint)
- may interfere with the current templating capabilities (re `{{ project }}` / `{{ group }}`).

Do you ( @gdubicki - project owner) have any plans (or even just opinions) on such functionality?

**Comment by gdubicki** [comment]
Hi @weakcamel , thank you for this report.

Frankly I have not thought about adding secure secrets handling to gitlabform yet.

At Egnyte we deploy gitlabform using Puppet, so we keep its config in a git repo, with all the values that are secret encrypted using asymmetric encryption - https://github.com/voxpupuli/hiera-eyaml. So on the machine which runs gitlabform once a day the config file is deployed in unencrypted form. This allows us to give developers access to the config repo and the public key so they can add MRs to add their new secrets encrypted. It is not perfect, but it's is safe enough for us thanks to other means of security that we use.

I think you can use hiera-eyaml even without using Puppet as it has cli to generate encryption keys, encrypt single value, edit encrypted file, decrypt single value, decrypt whole file - all the features to deploy the flow similar to ours. There is also https://github.com/mozilla/sops that seems to work in a similar way and has more features, but I have not use it myself.

Of course, as always, any PRs to gitlabform adding securing of the secrets into it are welcome too!

**Comment by weakcamel** [comment]
For your information (and for the record): for now I've worked around that with Jinja2 templates.
It's not quite ideal, however using [vault-cli](https://github.com/peopledoc/vault-cli) helped me to very easily fill in the secrets directly from Hashicorp Vault.

Once https://github.com/peopledoc/vault-cli/issues/113 is finished, I'm going to also break down the `config.yml` into smaller bits as I expect we may end up with quite a sizeable configuration.

I may still look at adding secret support (or maybe Jinja2 support?) to Gitlabform at some point, but probably not just yet.

**Comment by Dridge** [comment]
I had the same problem, ended up just using sed in the gitlab-ci.yml to replace like so:
`sed -i -e 's/SED_REPLACE_JIRA_PASSWORD/'"$JIRA_PASSWORD"'/g' config.yml`

**Comment by gdubicki** [comment]
How about adding support for automatic replacement of some strings with environment variables + changing logging code to not print any values with verbosity level up to „verbose” (they would still be printed with „debug” on though).

This would allow setting them as secrets in GitLab project, setting them from Vault in a pre-script etc. and seems easy to implement.

What do you think?

**Comment by gdubicki** [comment]
...or we could add built-in Jinja2 support to automatically pre-process the while config with built-in code similar to answers for https://stackoverflow.com/q/25862071/2693875 to be able to reference env variables.

**Comment by gdubicki** [comment]
...and this could remove uglyness of group, project and usernames used in the integration tests being both hardcoded in provided configs as well as in variables for all the rest of the code. With Jinja2 and env variables support we could just set the variables as env variables and then reference them in the configs. We could even randomize thise entities names to make the tests run on new entities each time and not run into problems with GitLab having a delay with actually deleting resources, to which you run into when you have static resource names. 😍

**Comment by weakcamel** [comment]
Yes, an ability to grab values from environment or to pass arbitrary Jinja variables on via CLI to gitlabform (e. g. `-e foo=bar -e baz=qux`) and so on would be great! The option with environment vars would probably be  more useful if someone wanted to make the logs public (would not expose the secrets in command-line parameters). 

The change in logging also sounds like a reasonable, simple way to hide the secrets.

**Comment by gdubicki** [comment]
Using Jinja2 templating everywhere would also be a nice workaround for YAML limitation: it can have anchor/alias for a mapping/object, but not for a sequence.

We would just define the sequence as a variable and then put it into places where we want it using Jinga2 tags.

**Comment by codrcodz** [comment]
We wrapped a few tools together into a CLI and threw it into a docker container:

- https://gitlab.com/dreamer-labs/repoman/repoman

The tools include:
- `gitlabform` for managing the configuration of existing repositories
- `python-gitlab` for creating and deleting repositories and groups
- `ansible-vault` for managing secrets

MRs welcome!

**Comment by saz** [comment]
We're executing the following script before running gitlabform in a pipeline:
```python3
#!/usr/bin/env python3
import os
from pathlib import Path
from string import Template

config_file = Path("./config.yml")
config_file.write_text(Template(config_file.read_text()).substitute(os.environ))
```

**Comment by amimas** [comment]
There are different tools/workflows/platforms exist just for secret management/handling. They specialize in that domain. In my opinion it should be okay to utilize other process to generate final input/config needed by gitlabform.

Just to add to existing suggestions mentioned above, here's another tool. Checkout [spruce](https://github.com/geofffranks/spruce), if you haven't already. It may not be immediately clear how it can be helpful for handling secrets. Checkout all the ["operators" supported by the tool](https://github.com/geofffranks/spruce/blob/main/doc/operators.md). There are operators for using environment variable in your config file. If you're using vault or aws secrets, there are first class support for that too. Some of the other operators might help you with other use cases.

I don't think handling secrets natively in gitlabform will be a high priority. There is very little resource maintaining gitlabform. It can't even keep up with regular features of gitlab and frankly needs improvements in other existing areas.

**Comment by TimKnight01** [comment]
Agreed - I think this is best handled by tools which can inject/replace Secrets into the YAML prior to being run, rather than having GitLabForm try and do YAML processing itself to read information from whatever environment the secrets might be held in. My gut feeling is it'd be relatively easy to do for Secrets which exist on the environment of the runner used to execute GitLabForm, but then we end up with a weird one use case scenario
